internal/controller/pkg/revision/establisher.go

```diff
@@ -108,136 +115,165 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
-					conf.Spec.Conversion.Webhook.ClientConfig.Service = &extv1.ServiceReference{}
+
+				// If resource does not already exist, we must attempt to dry run create
+				// it.
+				if kerrors.IsNotFound(err) {
+					// Add to objects as not existing.
+					allObjs <- currentDesired{
+						Desired: d,
+						Current: nil,
+						Exists:  false,
+					}
+					// We will not create a resource if we are not going to control it,
+					// so we don't need to check with dry run.
+					if control {
+						if err := e.create(ctx, d, parent, client.DryRunAll); err != nil {
+							return err
+						}
+					}
+					return nil
 				}
-				conf.Spec.Conversion.Webhook.ClientConfig.CABundle = webhookTLSCert
-				conf.Spec.Conversion.Webhook.ClientConfig.Service.Name = parent.GetName()
-				conf.Spec.Conversion.Webhook.ClientConfig.Service.Namespace = e.namespace
-				conf.Spec.Conversion.Webhook.ClientConfig.Service.Port = pointer.Int32(webhookPort)
-			}
-		}
 
-		// Make a copy of the desired object to be populated with existing
-		// object, if it exists.
-		copy := res.DeepCopyObject()
-		current, ok := copy.(client.Object)
-		if !ok {
-			return nil, errors.New(errAssertClientObj)
-		}
-		err := e.client.Get(ctx, types.NamespacedName{Name: d.GetName(), Namespace: d.GetNamespace()}, current)
-		if resource.IgnoreNotFound(err) != nil {
-			return nil, err
-		}
+				c := current.(resource.Object)
+				// Add to objects as existing.
+				allObjs <- currentDesired{
+					Desired: d,
+					Current: c,
+					Exists:  true,
+				}
 
-		// If resource does not already exist, we must attempt to dry run create
-		// it.
-		if kerrors.IsNotFound(err) {
-			// Add to objects as not existing.
-			allObjs = append(allObjs, currentDesired{
-				Desired: d,
-				Current: nil,
-				Exists:  false,
-			})
-			// We will not create a resource if we are not going to control it,
-			// so we don't need to check with dry run.
-			if control {
-				if err := e.create(ctx, d, parent, client.DryRunAll); err != nil {
-					return nil, err
+				if err := e.update(ctx, c, d, parent, control, client.DryRunAll); err != nil {
+					return err
 				}
-			}
-			continue
+				return nil
+			})
 		}
-
-		c := current.(resource.Object)
-		// Add to objects as existing.
-		allObjs = append(allObjs, currentDesired{
-			Desired: d,
-			Current: c,
-			Exists:  true,
-		})
-
-		if err := e.update(ctx, c, d, parent, control, client.DryRunAll); err != nil {
+		if err := g.Wait(); err != nil {
 			return nil, err
 		}
 	}
+	close(allObjs)
 
-	for _, cd := range allObjs {
-		if !cd.Exists {
-			// Only create a missing resource if we are going to control it.
-			// This prevents an inactive revision from racing to create a
-			// resource before an active revision of the same parent.
-			if control {
-				if err := e.create(ctx, cd.Desired, parent); err != nil {
-					return nil, err
+	refsCh := make(chan xpv1.TypedReference, len(allObjs))
```

Comment:
```suggestion
	refsCh := make(chan xpv1.TypedReference, maxConcurrentReconciles)
```

Similar to my comment above, would this make sense?